src/chainstate/coordinator/mod.rs

```diff
@@ -658,6 +708,281 @@ impl<
+        // read STX contract state
+        let stacks_tip = SortitionDB::get_block_snapshot_consensus(
+            self.sortition_db.conn(),
+            curr_block_consensus_hash,
+        )?
+        .ok_or(Error::ChainstateError(ChainstateError::NoSuchBlockError))?;
+
+        // ensure that PoX is active
+        let is_pox_active = self.sortition_db.is_pox_active_in_reward_cycle(
+            rc_cycle_of_vote,
+            &self.burnchain,
+            &stacks_tip,
+        )?;
+        if !is_pox_active {
+            return Ok(None);
+        }
+
+        let stacked_stx = self.chain_state_db.get_total_ustx_stacked(
+            &self.sortition_db,
+            &stacks_block_id,
+            rc_cycle_of_vote as u128,
+        )?;
+        // Want to round up here, so calculating (x + n - 1) / n here instead of (x / n)
+        let min_stx_for_valid_vote = ((stacked_stx
+            * self
+                .burnchain
+                .exit_contract_constants
+                .percent_stacked_stx_for_valid_vote as u128)
+            + 99)
+            / 100;
+
+        // obtain map of exit proposal to number of valid votes
+        let (vote_map, total_votes) = self.read_vote_state(
+            rc_cycle_of_vote,
+            curr_exit_at_rc_opt,
+            curr_block_consensus_hash,
+            curr_block_hash,
+            invalid_reward_cycles,
+            exit_contract_id,
+        )?;
+
+        // ensure that there are enough votes for a valid vote
+        if total_votes < min_stx_for_valid_vote as u128 {
+            return Ok(None);
+        }
+
+        // Explanation of voting mechanics: a vote to exit at RC x is a vote for the blockchain to
+        // exit at RC x OR higher. To count the votes, we iterate from the lowest to the highest
+        // valid exit proposal in the vote map, until the total accrued votes surpasses the
+        // threshold for consensus.
+        let min_stx_for_consensus = (min_stx_for_valid_vote
+            * self
+                .burnchain
+                .exit_contract_constants
+                .vote_confirmation_percent as u128
+            + 99)
+            / 100;
+        let mut accrued_votes = 0;
+        // Since vote map is a BTreeMap, iteration over keys will occur in a sorted order
+        for (curr_rc_proposal, curr_votes) in vote_map.iter() {
+            accrued_votes += curr_votes;
+            if accrued_votes > min_stx_for_consensus {
+                // If the accrued votes is greater than the minimum needed to achieve consensus,
+                // store this value for the upcoming veto
+                return Ok(Some(*curr_rc_proposal));
+            }
+        }
+        Ok(None)
+    }
+
+    pub fn reached_exit_reward_cycle(&self, header: &BurnchainBlockHeader) -> Result<bool, Error> {
+        if let Some(chain_tip) = self.canonical_chain_tip {
+            let exit_info_opt =
```

Comment:
This is only `Some(..)` once per reward cycle, right?

Reply:
No, I store a `BlockExitRewardCycleInfo` for every block in the Stacks chain.

Reply:
Updated the logic so that I only create this object once per reward cycle.